### PR TITLE
pipeline: refactor pipeline component

### PR DESCRIPTION
**Pull request by tlauda**
Refactors the whole implementation of pipeline component.

### CODE REVIEW

File: src/audio/pipeline.c [hunk 1]
```diff
@@ -247,8 +247,6 @@ struct pipeline *pipeline_new(struct sof_ipc_pipe_new *pipe_desc,
 	schedule_task_init(&p->pipe_task, pipeline_task, p);
 	schedule_task_config(&p->pipe_task, pipe_desc->priority,
 		pipe_desc->core);
-	list_init(&p->comp_list);
```
Comment (lgirdwood) [inline src/audio/pipeline.c]: Can you add some more detail in the commit message stating what APIs are duplicated and presumably that this patch merges these into a single API ?

**Comment by tlauda** [comment]
@lgirdwood This PR is supposed to refactor implementation of the pipeline component to simplify the graph traversal. I wanted to shrink the code by deduplicating upstream/downstream methods and just use the generic traversal function, since I've noticed it is easily doable. This PR shouldn't change the behavior of the methods, so if it does, it's my mistake. I will prepare further smaller PRs, which are going to change some of the functionalities and enable pipeline chaining, but that's the future.

File: src/audio/buffer.c [hunk 2]
```diff
@@ -85,7 +85,6 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
 	buffer->end_addr = buffer->addr + buffer->ipc_buffer.size;
 	buffer->free = buffer->ipc_buffer.size;
 	buffer->avail = 0;
-	buffer->connected = 0;
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/buffer.c]: Mixer is working based on the connected components and their state and it never uses buffer->connected parameter. My thinking here is that there is no point in keeping additional parameter, if it's only used for graph traversal and after my changes is used only in one place.

File: src/audio/pipeline.c [hunk 3]
```diff
@@ -74,8 +74,6 @@ static void connect_upstream(struct pipeline *p, struct comp_dev *start,
 
 	/* we are an endpoint if we have 0 source components */
 	if (list_is_empty(&current->bsource_list)) {
-		current->is_endpoint = 1;
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/pipeline.c]: Will update the commit message. It's basically the same as my intention for buffer->connected. No need to keep additional parameters if they are only used basically in one place.

File: src/audio/pipeline.c (hunk 1 above)
Comment (tlauda) [reply src/audio/pipeline.c]: Will do.

**Comment by lgirdwood** [comment]
@tlauda ok, lets get the validation done and then we can merge it.

**Comment by tlauda** [comment]
@mengdonglin What is the ETA for checking this PR?

**Comment by lgirdwood** [comment]
@tlauda fwiw, I will hold off other pipeline PRs until this is merged (unless critical fixes).

**Comment by lgirdwood** [comment]
@tlauda apologies, I had to merge the bigger checkpatch fix.

File: src/audio/pipeline.c [hunk 4]
```diff
@@ -206,57 +79,110 @@ struct pipeline *pipeline_new(struct sof_ipc_pipe_new *pipe_desc,
...
+	return 0;
+}
+
+/* generic method for walking the graph upstream or downstream */
+static int pipeline_for_each_comp(struct comp_dev *current,
+				  int (*func)(struct comp_dev *, void *, int),
+				  void *data,
+				  void (*buff_func)(struct comp_buffer *),
+				  int dir)
+{
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/pipeline.c]: It provides such functionality based on provided dir value.
I've introduced proper macros to make deduplication you are mentioning.

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/pipeline.c]: ok, but I dont see how this code will work when we have components that can split or merge streams like mixers, muxes, demuxers etc . e.g.. we need this function to recurse

**Comment by lgirdwood** [comment]
@tlauda could you rebase and force push (after CI is back and we will merge today).

File: src/audio/pipeline.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/pipeline.c]: ok, now I see the recursion by passing the function pointer. Can you add this in the func comments (just so it's obvious to others too). Thanks

**Comment by lgirdwood** [comment]
@tlauda ping

**Comment by keyonjie** [comment]
@tlauda @lgirdwood here it is:
https://github.com/thesofproject/sof/pull/900

**Comment by lgirdwood** [comment]
@keyonjie thanks, but I need you to describe the regression in #900 commit message

**Pull request merged**